cmake/modules/extensions.cmake

```diff
@@ -2507,76 +2538,61 @@ function(zephyr_get variable)
     set(GET_VAR_SYSBUILD "GLOBAL")
   endif()
 
-  if(GET_VAR_MERGE)
-    # Clear variable before appending items in MERGE mode but keep a backup for
-    # local appending later
-    set(local_var_backup ${${variable}})
-    set(${variable})
+  if(NOT DEFINED GET_VAR_VAR)
+    set(GET_VAR_VAR ${variable})
   endif()
 
-  set(used_global false)
-
-  if(SYSBUILD)
-    get_property(sysbuild_name TARGET sysbuild_cache PROPERTY SYSBUILD_NAME)
-    get_property(sysbuild_main_app TARGET sysbuild_cache PROPERTY SYSBUILD_MAIN_APP)
-    get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${sysbuild_name}_${variable})
-    if(NOT DEFINED sysbuild_${variable} AND
-       ("${GET_VAR_SYSBUILD}" STREQUAL "GLOBAL" OR sysbuild_main_app)
-    )
-      get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
-      set(used_global true)
-    endif()
-  endif()
+  # Keep current scope variables in internal variables.
+  # This is needed to properly handle cases where we want to check value against
+  # environment value or when appending with the MERGE operation.
+  foreach(var ${GET_VAR_VAR})
+    set(current_${var} ${${var}})
+    set(${var})
 
-  if(DEFINED sysbuild_${variable})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} ${sysbuild_${variable}})
-    else()
-      set(${variable} ${sysbuild_${variable}} PARENT_SCOPE)
-      return()
-    endif()
-  endif()
-  if(SYSBUILD AND GET_VAR_MERGE AND NOT used_global AND "${GET_VAR_SYSBUILD}" STREQUAL "GLOBAL")
-    get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
-    list(APPEND ${variable} ${sysbuild_${variable}})
-  endif()
-  if(DEFINED CACHE{${variable}})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} $CACHE{${variable}})
-    else()
-      set(${variable} $CACHE{${variable}} PARENT_SCOPE)
-      return()
-    endif()
-  endif()
-  if(DEFINED ENV{${variable}})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} $ENV{${variable}}})
+    if(SYSBUILD)
+      get_property(sysbuild_name TARGET sysbuild_cache PROPERTY SYSBUILD_NAME)
+      get_property(sysbuild_main_app TARGET sysbuild_cache PROPERTY SYSBUILD_MAIN_APP)
```

Comment:
This doesn't need to be ran multiple times in the for() loop

Reply:
correct, but I preferred keeping those lines close to the line below:
https://github.com/zephyrproject-rtos/zephyr/blob/53c2430dbe2a4c2ee0a2c536f2ae744ad7dbee5f/cmake/modules/extensions.cmake#L2555

if moving those two lines outside the for-loop then there would be a need one more `if(SYSBUILD) ... endif()` guard.
I consider the current approach more readable where the sysbuild code is located together.